### ISSUE TITLE
[java] LawOfDemeter: False positive with fields assigned to local vars

### ISSUE THREAD

**Issue by Fernal73** [pmd/pmd#2188]
https://github.com/pmd/pmd/issues/2160#issuecomment-565639733
**Affects PMD Version:** 
6.20.0.

**Rule:**
LawOfDemeter.

**Description:**
Assigning an instance or local variables (created locally) should not be flagged as an object created elsewhere by this rule.

**Code Sample demonstrating the issue:**

```java
package pmdtests;

public class AddressUsingCopyOnWrite {
  private volatile AddressValue addressValue;

 public AddressUsingCopyOnWrite(String       street, String city, String phone) {
   this.addressValue = new                    AddressValue(street, city, phone);
   }
  

 // thread-safe method
 @Override
  public String toString() {
    AddressValue local = addressValue;
    return "street=" + local.getStreet() + ",city=" + local.getCity()
        + ",phoneNumber=" + local.getPhoneNumber();
  }

 
}

```

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

**Comment by adangel** [comment]
Not sure, if I understand this correctly.
Here's an example how a locally created variable would look: `AdressValue local = new AddressValue("foo", "bar", ...)`

In your code snippet, we assign a field to a local variable - the field is our own, that's why we should not flag access to `local`. Is this correct?

So, it seems, that we don't consider fields when checking local variable initializers (I think, we check for method parameters, though).

**Comment by Fernal73** [comment]
(This would help if you came back to me within a shorter period of the issue written up. It takes some time to recall why it was written up in the first place. )

I'd expect the field assignment was to a locally created instance, not created elsewhere as specified by the rule. Perhaps, the code example ought to have examples of fields initialised in both ways to check whether that's the case. Am I making sense?

**Comment by jsotuyod** [comment]
This is fixed in PMD 7.0.0-rc1. It won't be backported to PMD 6.
